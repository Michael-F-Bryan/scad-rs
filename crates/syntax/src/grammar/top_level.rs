use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
