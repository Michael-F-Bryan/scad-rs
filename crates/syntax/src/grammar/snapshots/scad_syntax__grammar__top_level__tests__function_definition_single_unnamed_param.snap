---
source: crates/syntax/src/grammar/top_level.rs
description: statement(function x(y) = y + 2;)
---
Node(
    NAMED_FUNCTION_DEFINITION@0..22
      FUNCTION_KW@0..8 "function"
      WHITESPACE@8..9 " "
      IDENT@9..10 "x"
      L_PAREN@10..11 "("
      PARAMETERS@11..12
        IDENT@11..12
          IDENT@11..12 "y"
      R_PAREN@12..13 ")"
      WHITESPACE@13..14 " "
      EQUALS@14..15 "="
      BIN_EXPR@15..21
        LOOKUP_EXPR@15..17
          WHITESPACE@15..16 " "
          IDENT@16..17 "y"
        PLUS@17..19
          WHITESPACE@17..18 " "
          PLUS@18..19 "+"
        INTEGER@19..21
          WHITESPACE@19..20 " "
          INTEGER@20..21 "2"
      SEMICOLON@21..22 ";"
    ,
)
